File: src/main/java/com/leetcode/algorithm/wait/DominoAndTrominoTiling.java
package com.leetcode.algorithm.wait;

/**
 * DP1 输入是 O(n)
 * dp(x) 依赖于有限个解
 * <p>
 * 两种牌覆盖问题：现有两种类型的牌，每一种类型都可以进行旋转如下：
 * XX  ==> X    XX  ==> XX  ==> X    ==>  X
 * X    X        X      XX       XX
 * 限制：铺满，即在 2* N 的区域内，全覆盖，即 2*N 的区域内全是 X
 * 要求: 给出有多少种可能性
 * <p>
 * 解法：https://www.youtube.com/watch?v=S-fUTfqrdq8
 *
 * @author wezhyn
 * @since 03.15.2020
 */
public class DominoAndTrominoTiling {

    public int numTilings(int n) {
        int[][] dp=new int[n][2];
//         f(x)=dp[i-1][0]+dp[i-2][0] + dp[i-1][1]+ dp[i-1][2]
//        dp[i-1][1]=dp[i-2][0]+dp[i-1][2]
        return 0;
/*        dp[0][0]=dp[1][0]=1;
        for (int i=2; i <=n; i++) {
            dp[i][0]=

        }*/
    }

}

File: leetcode-kotlin/src/main/kotlin/com/wezhyn/learn/dp/limit/DominoAndTrominoTiling.kt
package com.wezhyn.learn.dp.limit

/**
 * 可以得知 dp(i)[0] 表示当前(i)所在的是 一字形
 * dp(i)[1] 是凸字形
 * @author wezhyn
 * @since 07.24.2020
 *
 */
class DominoAndTrominoTiling {
    private lateinit var cache: Array<IntArray>
    fun numTilings(N: Int): Int {
        cache = Array(N + 2) { IntArray(2) { 0 } }
        cache[1][0] = 1
        cache[2][0] = 2
        cache[2][1] = 1
        return dp(N, 0)
    }

    private fun dp(i: Int, y: Int): Int {
        var num = cache[i][y].toLong()
        if ((i > 2 && num != 0L) || i <= 2) {
            return num.toInt()
        }
        when (y) {
            0 -> {
                num = dp(i - 1, 0).toLong() + dp(i - 2, 0) + 2 * dp(i - 1, 1)
            }
            1 -> {
                num = dp(i - 1, 1).toLong() + dp(i - 2, 0)
            }
        }
        num = num.rem(1000000007)
        cache[i][y] = num.toInt()
        return cache[i][y]
    }
}